crates/red_knot/src/salsa_db/semantic/module.rs

```diff
@@ -0,0 +1,886 @@
+    Package,
+}
+
+/// A search path in which to search modules.
+/// Corresponds to a path in [`sys.path`](https://docs.python.org/3/library/sys_path_init.html) at runtime.
+///
+/// Cloning a search path is cheap because it's an `Arc`.
+#[derive(Clone, PartialEq, Eq)]
+pub struct ModuleSearchPath {
+    inner: Arc<ModuleSearchPathInner>,
+}
+
+impl ModuleSearchPath {
+    #[allow(unused)]
+    pub fn new(path: PathBuf, kind: ModuleSearchPathKind) -> Self {
+        Self {
+            inner: Arc::new(ModuleSearchPathInner { path, kind }),
+        }
+    }
+
+    #[allow(unused)]
+    pub fn kind(&self) -> ModuleSearchPathKind {
+        self.inner.kind
+    }
+
+    pub fn path(&self) -> &Path {
+        &self.inner.path
+    }
+}
+
+impl std::fmt::Debug for ModuleSearchPath {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct ModuleSearchPathInner {
+    path: PathBuf,
+    kind: ModuleSearchPathKind,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum ModuleSearchPathKind {
+    // Project dependency
+    FirstParty,
+
+    // e.g. site packages
+    ThirdParty,
+
+    // e.g. built-in modules, typeshed
+    #[allow(unused)]
+    StandardLibrary,
+}
+
+impl ModuleSearchPathKind {
+    #[allow(unused)]
+    pub const fn is_first_party(self) -> bool {
+        matches!(self, Self::FirstParty)
+    }
+}
+/// ID uniquely identifying a module.
+#[salsa::input(jar=Jar, singleton)]
+pub struct ModuleSearchPaths {
+    #[return_ref]
+    pub paths: Vec<ModuleSearchPath>,
+}
+
+//////////////////////////////////////////////////////
+// Queries
+//////////////////////////////////////////////////////
+
+pub fn module_search_paths(db: &dyn Db) -> &[ModuleSearchPath] {
+    ModuleSearchPaths::get(db).paths(db).as_slice()
+}
+
+pub fn resolve_module_name(db: &dyn Db, name: ModuleName) -> Option<ResolvedModule> {
+    resolve_module(db, Module::new(db, name))
+}
+
+/// Resolves a module name to a module id
+#[tracing::instrument(level = "debug", skip(db))]
+#[salsa::tracked(jar=Jar)]
+pub fn resolve_module(db: &dyn Db, module: Module) -> Option<ResolvedModule> {
+    let name = module.name(db);
+
+    let Some((root_path, resolved_file, kind)) = resolve_module_path(db, &name) else {
+        return None;
+    };
+
+    let normalized = resolved_file
+        .path(db.upcast())
+        .canonicalize()
+        .map(|path| db.file(path))
+        .unwrap_or_else(|_| resolved_file);
+
+    Some(ResolvedModule::new(
+        db,
+        name.clone(),
+        kind,
+        root_path,
+        normalized,
+    ))
+}
+
+/// Resolves the module id for the given path.
+///
+/// Returns `None` if the path is not a module in `sys.path`.
+#[tracing::instrument(level = "debug", skip(db))]
+pub fn path_to_module(db: &dyn Db, path: &Path) -> Option<ResolvedModule> {
+    let file = db.file(path.to_path_buf());
```

Comment:
It's a bit weird that `path_to_module` converts the path to a `file` as the very first thing only so that `file_to_module` then reads the path. However, for `file_to_module` to be a salsa query, it can only accept an ingredient as an argument and `file` is an ingredient but `path` isn't.